# 错误处理和资源管理

## defer调用

`````go
package main

import "fmt"

func tryDefer() {
	// 先进后出
	defer fmt.Println(1)  //出错依然会处理
	defer fmt.Println(2)
	panic("error")
	fmt.Println(3)
}

func main() {
	tryDefer()
}
`````

````go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strings"
)

type intGen func() int

func (g intGen) Read(p []byte) (n int, err error) {
	next := g()
	if next > 10000{
		return 0, io.EOF
	}
	s := fmt.Sprintf("%d\n", next)
	return strings.NewReader(s).Read(p)
}

func fibonacci() intGen {
	a, b := 0, 1
	return func() int {
		a, b = b, a+b
		return a
	}
}


func writeFile(filename string){
	file, err := os.Create(filename)
	if err != nil{
		panic(err)
	}
	defer file.Close()

	f := fibonacci()
	writer := bufio.NewWriter(file) // 这样会先写入内存之后一次性写入，速度快
	defer writer.Flush() //执行写入操作

	for i := 0; i< 500; i++{
		fmt.Fprintln(writer, f())
	}
}


func main() {
	writeFile("a.txt")
}
````

## 错误处理

`````go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"io"
	"os"
	"strings"
)

type intGen func() int

func (g intGen) Read(p []byte) (n int, err error) {
	next := g()
	if next > 10000 {
		return 0, io.EOF
	}
	s := fmt.Sprintf("%d\n", next)
	return strings.NewReader(s).Read(p)
}

func fibonacci() intGen {
	a, b := 0, 1
	return func() int {
		a, b = b, a+b
		return a
	}
}

func writeFile(filename string) {
	file, err := os.OpenFile(filename, os.O_EXCL|os.O_CREATE, 0666)

	err = errors.New("this is a custom error") // 自己定义一个error
	if err != nil {
		if pathError, ok := err.(*os.PathError); !ok { // 不是预期的错误
			panic(err)
		} else {
			fmt.Printf("%s, %s, %s\n", pathError.Op, pathError.Path, pathError.Err)
		}
		//fmt.Println("Error:", err)
		return
	}
	defer file.Close()

	f := fibonacci()
	writer := bufio.NewWriter(file) // 这样会先写入内存之后一次性写入，速度快
	defer writer.Flush()            //执行写入操作

	for i := 0; i < 500; i++ {
		fmt.Fprintln(writer, f())
	}
}

func main() {
	writeFile("a.txt")
}
`````

## 服务器统一出错处理

`````
└─filestringserver
    └─filelisting
   		|_handler.go
   	|_web.go
`````

`````go
package main

import (
	"go_test/filestringserver/filelisting"
	"net/http"
	"os"
	"log"
)

type appHandler func(w http.ResponseWriter, r *http.Request) error

func errWrapper(handler appHandler) func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		err := handler(w, r)
		if err != nil {
			log.Printf(err.Error())
			code := http.StatusOK
			switch {
			case os.IsNotExist(err):
				//http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
				code = http.StatusNotFound
			case os.IsPermission(err):
				code = http.StatusForbidden
			default:
				code = http.StatusInternalServerError
			}
			http.Error(w, http.StatusText(code), code)

		}
	}
}

func main() {
	http.HandleFunc("/list/", errWrapper(filelisting.Handle))

	err := http.ListenAndServe(":8888", nil)
	if err != nil {
		panic(err)
	}
}
`````

```go
package filelisting

import (
	"io/ioutil"
	"net/http"
	"os"
)

func Handle(writer http.ResponseWriter, request *http.Request) error {
	path := request.URL.Path[len("/list/"):]
	file, err := os.Open(path)
	if err != nil {
		//http.Error(writer, err.Error(), http.StatusInternalServerError)
		//return

		return err
	}
	defer file.Close()

	all, err := ioutil.ReadAll(file)
	if err != nil {
		return err
	}
	writer.Write(all)

	return  nil
}
```

## panic和recover

### panic

* 停止当前函数的执行
* 一直向上返回，执行每一层的defer
* 如果没有遇到recover，程序退出

### recover

* 仅仅在defer调用中使用
* 获取panic的值
* 无法处理，可重新painc

````go
package main

import (
	"fmt"
)

func tryRecover() {
	defer func() {
		r := recover()
		if err, ok := r.(error); ok { //判断类型是error
			fmt.Println("Error occurred:", err)
		} else { // 不是error不确定是什么东西
			panic(r)
		}
	}()
	//panic(errors.New("this is an error"))

	//b := 0
	//a := 4 / b
	//fmt.Println(a)

	panic("123")  // recover会发现 这个东西 不是error
}

func main() {
	tryRecover()
}
````

##  服务器统一出错处理2

`````go
//web.go

package main

import (
	"go_test/filestringserver/filelisting"
	"log"
	"net/http"
	"os"
)

type appHandler func(w http.ResponseWriter, r *http.Request) error

func errWrapper(handler appHandler) func(http.ResponseWriter, *http.Request) {
	return func(w http.ResponseWriter, r *http.Request) {
		defer func() {
			if r := recover(); r != nil {
				log.Printf("Panic: %v", r)
				http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
			}
		}()

		err := handler(w, r)
		if err != nil {
			log.Printf(err.Error())

			if userErr, ok := err.(userError); ok{
				http.Error(w, userErr.Message(), http.StatusBadRequest)
				return
			}
			code := http.StatusOK
			switch {
			case os.IsNotExist(err):
				//http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
				code = http.StatusNotFound
			case os.IsPermission(err):
				code = http.StatusForbidden
			default:
				code = http.StatusInternalServerError
			}
			http.Error(w, http.StatusText(code), code)

		}
	}
}

type userError interface{
	error
	Message() string
}

func main() {
	http.HandleFunc("/list/", errWrapper(filelisting.Handle))

	err := http.ListenAndServe(":8888", nil)
	if err != nil {
		panic(err)
	}
}
`````

`````go
//handler.go

package filelisting

import (
	"io/ioutil"
	"net/http"
	"os"
	"strings"
)

const prefix = "/list/"


type userError string


func (e userError) Error() string {
	return e.Message()
}

func (e userError) Message() string {
	return string(e)
}

func Handle(writer http.ResponseWriter, request *http.Request) error {
	if strings.Index(request.URL.Path, prefix) != 0 {
		return userError("PATH must start with " + prefix)
	}
	path := request.URL.Path[len(prefix):]
	file, err := os.Open(path)
	if err != nil {
		//http.Error(writer, err.Error(), http.StatusInternalServerError)
		//return

		return err
	}
	defer file.Close()

	all, err := ioutil.ReadAll(file)
	if err != nil {
		return err
	}
	writer.Write(all)

	return nil
}
`````
